### PR TITLE
Transform meta for codecs

### LINKED ISSUE #67: Codecs should transform meta
I.e. AudioCodec: when changing sampling rate, it should also change samples count.
For VideoCodec frame rate could be changed and frames count (#66) will also change.

### PR DESCRIPTION
**Pull request by krkmetal**
Closes #67

### CODE REVIEW

File: fffw/encoding/outputs.py [hunk 1]
```diff
@@ -93,6 +93,17 @@ def check_buffering(self) -> Optional[List[str]]:
...
 
+    @property
+    def meta(self) -> Optional[Meta]:
+        metadata = self.get_meta_data(self)
+        if metadata is None:
+            return None
+        return self.transform(metadata)
+
+    def transform(self, metadata: Meta) -> Meta:
+        """ Apply filter changes to codec metadata."""
```
Comment (tumb1er) [inline fffw/encoding/outputs.py:104]: "Apply codec changes to stream metadata."

File: fffw/encoding/outputs.py [hunk 2]
```diff
@@ -93,6 +93,17 @@ def check_buffering(self) -> Optional[List[str]]:
             prev = scene
         return meta.streams
 
+    @property
+    def meta(self) -> Optional[Meta]:
```
Comment (tumb1er) [inline fffw/encoding/outputs.py:97]: base `transform` method for filters is defined in `Node` class. For codecs corresponding base class is `Dest`, there things should be placed there.

File: tests/test_vector.py [hunk 3]
```diff
@@ -71,9 +80,15 @@ def test_vector_metadata(self):
             self.assertEqual(v.meta, expected)
 
         with self.subTest("codec meta"):
-            simd = SIMD(self.source, self.output1)
-            x = v > simd
-            self.assertEqual(x.meta, expected)
+            a = self.simd.audio
```
Comment (tumb1er) [inline tests/test_vector.py:83]: This is a smoke test for `Vector.meta` property. 
* look at `FilterGraphTestCase`
* there is no need to add extra output to `SIMD` here, just add bitrate parameter in `setUp`

**Review by tumb1er — CHANGES_REQUESTED** [review]

**Comment by tumb1er** [comment]
Diff coverage is incomplete.

File: fffw/encoding/outputs.py (hunk 1 above)
Comment (krkmetal) [reply fffw/encoding/outputs.py]: fixed

File: fffw/encoding/outputs.py (hunk 2 above)
Comment (krkmetal) [reply fffw/encoding/outputs.py]: Moved `transform` and `meta` to `Dest` class.

File: tests/test_vector.py (hunk 3 above)
Comment (krkmetal) [reply tests/test_vector.py]: Updated `vector_metadata` test and added tests to `FilterGraphTestCase` including some cases to fix coverage.

File: fffw/encoding/outputs.py [hunk 4]
```diff
@@ -2,7 +2,7 @@
 from itertools import chain
 from typing import List, cast, Optional, Iterable, Any
 
-from fffw.graph.meta import AUDIO, VIDEO, StreamType
+from fffw.graph.meta import AUDIO, VIDEO, StreamType, Meta
```
Comment (tumb1er) [inline fffw/encoding/outputs.py:5]: Unused import?

File: fffw/graph/base.py [hunk 5]
```diff
@@ -59,7 +59,14 @@ def name(self) -> str:
 
     @property
     def meta(self) -> Optional[Meta]:
-        return self.get_meta_data(self)
+        metadata = self.get_meta_data(self)
+        if metadata is None:
+            return None
+        return self.transform(metadata)
+
+    def transform(self, *metadata: Meta) -> Meta:
```
Comment (tumb1er) [inline fffw/graph/base.py:67]: Codec accepts only single input stream. There is no need to support multiple, if there are no complains from MyPy.

File: tests/test_graph.py [hunk 6]
```diff
@@ -40,7 +50,10 @@ def setUp(self) -> None:
         self.audio_metadata = audio_meta_data(
             duration=200.0,
             sampling_rate=48000,
-            samples_count=200 * 48000)
+            samples_count=200 * 48000,
```
Comment (tumb1er) [inline tests/test_graph.py:53]: Is this 200 the same as previous 200.0 one? What about 48000?

**Review by tumb1er — CHANGES_REQUESTED** [review]

File: fffw/graph/base.py (hunk 5 above)
Comment (krkmetal) [reply fffw/graph/base.py:67]: Yes, MyPy found conflict:
```
Definition of "transform" in base class "Dest" is incompatible with definition in base class "Node"
```
I'm not sure why but Filter definition inherits both Node and Dest for TYPE_CHECKING:
```
class Filter(mixins.StreamValidationMixin, base.Node, Params)
...
class StreamValidationMixin(StreamValidationTarget)
...
if TYPE_CHECKING:
    StreamValidationTarget = base.Dest
else:
    StreamValidationTarget = object
```

File: fffw/encoding/outputs.py (hunk 4 above)
Comment (krkmetal) [reply fffw/encoding/outputs.py:5]: fixed

File: tests/test_graph.py (hunk 6 above)
Comment (krkmetal) [reply tests/test_graph.py]: I see no real difference here. Changed to `200.0 * 48000` for consistency. What's wrong with 48000 other than being hardcoded?

File: fffw/graph/base.py (hunk 5 above)
Comment (tumb1er) [reply fffw/graph/base.py:67]: Both Node and Dest have input edges, it's the main reason.

**Review by tumb1er — CHANGES_REQUESTED** [review]

File: tests/test_graph.py (hunk 6 above)
Comment (tumb1er) [reply tests/test_graph.py]: Magic numbers.

Comment (krkmetal) [reply tests/test_graph.py]: fixed

**Review by tumb1er — APPROVED** [review]

**Pull request merged**